[mlir][vector] Update tests for collapse 3/n (nfc)

banach-space opened the pull request:
The main goal of this PR (and subsequent PRs), is to add more tests with
scalable vectors to:
  * vector-transfer-collapse-inner-most-dims.mlir

There's quite a few cases to consider, hence this is split into multiple
PRs. In this PR, the very first test for `vector.transfer_write` is
complemented with all the possible combinations:
  * scalable (rather than fixed) unit trailing dim,
  * dynamic (rather than static) trailing dim in the source memref.

To this end, the following tests:
  * `@leading_scalable_dimension_transfer_write`
    `@trailing_scalable_one_dim_transfer_write`

are replaced with:
  * `@drop_two_inner_most_dim_scalable_inner_dim` and
    `@negative_scalable_unit_dim`,

respectively. In addition:
  * "_for_transfer_write" is removed from function names (to reduce
    noise).

In addition, to maintain consistency between the tests for `xfer_read`
and `xfer_write`, 2 negative tests for `xfer_read` are also renamed.
This is to follow the suggestion made during the review of this PR.

This is a follow-up for: #94490 and #94604

NOTE: This PR is limited to tests for `vector.transfer_write`.

banach-space committed:
[mlir][vector] Update tests for collapse 3/n (nfc)

The main goal of this PR (and subsequent PRs), is to add more tests with
scalable vectors to:
  * vector-transfer-collapse-inner-most-dims.mlir

There's quite a few cases to consider, hence this is split into multiple
PRs. In this PR, the very first test for `vector.transfer_write` is
complemented with all the possible combinations:
  * scalable (rather than fixed) unit trailing dim,
  * dynamic (rather than static) trailing dim in the source memref.

To this end, the following tests:
  * `@leading_scalable_dimension_transfer_write`
    `@trailing_scalable_one_dim_transfer_write`

are replaced with:
  * `@drop_two_inner_most_dim_scalable_inner_dim` and
    `@negative_scalable_unit_dim`,

respectively. In addition:
  * "_for_transfer_write" is removed from function names (to reduce
    noise).

This is a follow-up for: #94490, #94604

NOTE: This PR is limited to tests for `vector.transfer_write`.

MacDue commented on mlir/test/Dialect/Vector/vector-transfer-collapse-inner-most-dims.mlir:
```diff
@@ -272,16 +272,67 @@ func.func @drop_two_inner_most_dim_for_transfer_write(%arg0: memref<1x512x16x1x1
+// CHECK-NOT: memref.subview
+// CHECK-NOT: vector.shape_cast
+
+// Same as the top example within this split, but with a scalable unit dim in
+// the output vector - not supported
```
```suggestion
// Same as the top example within this split, but with a scalable one dim in
// the output vector - not allowed
```
There's no such thing as a scalable unit dim (for AArch64 `[1] ` can contain up to 16 elements!)

banach-space replied:
> There's no such thing as a scalable unit dim for AArch64 [1] can contain up to 16 elements!

This test is target-agnostic - the limitations of any specific target are irrelevant here. If we want to allow/disallow sth, then we need formalise and justify it (i.e. document it). I don't think that's necessary.

As for "scalable unit" vs "scalable one" for `[1]`, I don't follow. The very presence of "scalable" in the name highlights that this is something special. By using "unit" we are consistent with the other tests and the terminology used in the pattern definition. What's wrong with "unit"?

If we do advocate for inconsistency, then please formalise the difference between "scalable one" and "scalable unit" through documentation.

MacDue replied:
unit = an individual thing

But a scalable one dim is _not_ an individual thing, depending on vscale it can be up to 16 elements (for AArch64, but for any scalable target it won't always be one individual element).

MacDue replied:
So, I prefer scalable one dim as scalable unit dim feels like an oxymoron.

banach-space replied:
> unit = an individual thing

Indeed, and the meaning of "thing" is up for interpretation - yours is different to mine. 

While there's no *documented* definition of "scalable one"/"scalable unit", this is all about our individual preferences ...

> So, I prefer "scalable one dim"

Anyway, I've renamed these tests and added comments in "VectorTransforms.cpp".

MacDue commented on mlir/test/Dialect/Vector/vector-transfer-collapse-inner-most-dims.mlir:
```diff
@@ -272,16 +272,67 @@ func.func @drop_two_inner_most_dim_for_transfer_write(%arg0: memref<1x512x16x1x1
+// Same as the top example within this split, but with a scalable unit dim in
+// the output vector - not supported
+
+func.func @negative_scalable_unit_dim(%arg0: memref<1x512x16x1x1xf32>, %arg1: vector<1x16x16x1x[1]xf32>, %arg2: index) {
```
```suggestion
func.func @negative_scalable_one_dim(%arg0: memref<1x512x16x1x1xf32>, %arg1: vector<1x16x16x1x[1]xf32>, %arg2: index) {
```

banach-space replied:
"unit" is consistent with what's used in the pattern definition and the goal here is to remain consistent
https://github.com/llvm/llvm-project/blob/77db8b08c8b186c2625f8dfb26bb976561b43c4c/mlir/lib/Dialect/Vector/Transforms/VectorTransforms.cpp#L1338-L1354

MacDue replied:
The pattern is removing unit dims yes, but `[1]` is not a unit dim.

banach-space replied:
> it's something else the pattern should fail one

I will update the comments to better document the behaviour. For consistency with the docs and the existing names, I am keeping "scalable unit" (happy to keep in quotes to highlight that it's a special 🌷 ) . As pointed out in my other comment, "scalable unit" vs "scalable one" is a matter of individual preference unless we formalise the definition.

banach-space committed:
fixup! [mlir][vector] Update tests for collapse 3/n (nfc)

Rename tests, expand docs

banach-space committed:
fixup! fixup! [mlir][vector] Update tests for collapse 3/n (nfc)

Rename xfer_write rather than xfer_read tests. In the previous PR I renamed xfer_read tests instead. For consistency, I'll keep both changes and update the summary accordingly.
